cdap-watchdog/src/main/java/co/cask/cdap/logging/logbuffer/LogBufferRecoveryService.java

```diff
@@ -0,0 +1,154 @@
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import com.google.common.util.concurrent.MoreExecutors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Log buffer recovery service which recovers logs upon log saver restart and sends them to log buffer pipeline for
+ * further processing.
+ */
+class LogBufferRecoveryService extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(LogBufferRecoveryService.class);
+  // For outage, only log once per 60 seconds per message.
+  private static final Logger OUTAGE_LOG =
+    Loggers.sampling(LOG, LogSamplers.perMessage(() -> LogSamplers.limitRate(60000)));
+  private static final String SERVICE_NAME = "log.buffer.recovery";
+
+  private final List<LogBufferProcessorPipeline> pipelines;
+  private final List<CheckpointManager<LogBufferFileOffset>> checkpointManagers;
+  private final String baseLogDir;
+  private final int batchSize;
+  private final CountDownLatch stopLatch;
+
+  private LogBufferReader reader;
+  private volatile boolean stopped;
+
+  LogBufferRecoveryService(CConfiguration cConf, List<LogBufferProcessorPipeline> pipelines,
+                           List<CheckpointManager<LogBufferFileOffset>> checkpointManagers) {
+    this(pipelines, checkpointManagers, cConf.get(Constants.LogBuffer.LOG_BUFFER_BASE_DIR),
+         cConf.getInt(Constants.LogBuffer.LOG_BUFFER_RECOVERY_BATCH_SIZE));
+  }
+
+  @VisibleForTesting
+  LogBufferRecoveryService(List<LogBufferProcessorPipeline> pipelines,
+                           List<CheckpointManager<LogBufferFileOffset>> checkpointManager,
+                           String baseLogDir, int batchSize) {
+    this.pipelines = pipelines;
+    this.checkpointManagers = checkpointManager;
+    this.baseLogDir = baseLogDir;
+    this.batchSize = batchSize;
+    this.stopLatch = new CountDownLatch(1);
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    // get the smallest offset of all the log pipelines
+    LogBufferFileOffset minOffset = getSmallestOffset(checkpointManagers);
+    this.reader = new LogBufferReader(baseLogDir, batchSize, minOffset.getFileId(), minOffset.getFilePos());
+  }
+
+  @Override
+  protected void run() throws Exception {
+    // check if the log buffer dir exists, this could happen if log saver is starting for the first time.
+    if (!exists(baseLogDir)) {
+      return;
+    }
+
+    List<LogBufferEvent> logBufferEvents = new LinkedList<>();
+    boolean hasReadEvents = true;
+    while (!stopped && hasReadEvents) {
+      try {
+        hasReadEvents = reader.readEvents(logBufferEvents) > 0;
+        recoverLogs(logBufferEvents, pipelines);
```

Comment:
What happens if the recovery service crashes partway?

Reply:
If the recovery service crashes, the container should restart because without recovering old logs, new logs should not get processed.

Reply:
In that case, the recovery operation will be retried.